app/src/main/java/edu/wpi/first/shuffleboard/app/json/SourcedRestorer.java

```diff
@@ -0,0 +1,57 @@
+package edu.wpi.first.shuffleboard.app.json;
+
+import edu.wpi.first.shuffleboard.api.util.TypeUtils;
+import edu.wpi.first.shuffleboard.api.widget.Sourced;
+import edu.wpi.first.shuffleboard.app.sources.DestroyedSource;
+
+import java.util.Collection;
+import java.util.List;
+import java.util.function.BiConsumer;
+import java.util.stream.Collectors;
+
+/**
+ * Deals with destroyed sources and restoring those sources.
+ */
+public class SourcedRestorer {
+
+  /**
+   * Adds a {@link DestroyedSource} for each supported data type of a sourced object. The application will handle
+   * restoring the sources when possible.
+   *
+   * @param sourced   the sourced object to add destroyed sources to
+   * @param sourceUri the URI of the source
+   */
+  public void addDestroyedSourcesForAllDataTypes(Sourced sourced, String sourceUri) {
+    sourced.addSource(DestroyedSource.forUnknownData(sourced.getDataTypes(), sourceUri));
+  }
+
+  /**
+   * Attempts to restore all destroyed sources in a sourced object for the given URIs.
+   *
+   * @param sourced       the sourced to restore sources for
+   * @param urisToRestore the URIs of the sources to restore
+   * @param errorHandler  a function to call when a source could not be restored, or when a restored source could not
+   *                      be added
+   */
+  public void restoreSourcesFor(Sourced sourced,
+                                Collection<? extends String> urisToRestore,
+                                BiConsumer<DestroyedSource, ? super Throwable> errorHandler) {
```

Comment:
An error handler gets passed in so that sources can still be restored if an exception was thrown in the restoration loop. Right now all this does is log the error and say which source wasn't able to be restored.